File: crates/rspack_plugin_externals/src/electron_target_plugin.rs
use rspack_core::{BoxPlugin, ExternalItem, PluginExt};

use crate::ExternalsPlugin;

#[derive(Debug)]
pub enum ElectronTargetContext {
  Main,
  Preload,
  Renderer,
  None,
}

impl From<String> for ElectronTargetContext {
  fn from(value: String) -> Self {
    match value.as_str() {
      "main" => Self::Main,
      "preload" => Self::Preload,
      "renderer" => Self::Renderer,
      "none" => Self::None,
      _ => {
        unreachable!("ElectronTargetContext should only be one of main, preload, renderer, none")
      }
    }
  }
}

pub fn electron_target_plugin(context: &ElectronTargetContext, plugins: &mut Vec<BoxPlugin>) {
  plugins.push(
    ExternalsPlugin::new(
      "node-commonjs".to_string(),
      [
        "clipboard",
        "crash-reporter",
        "electron",
        "ipc",
        "native-image",
        "original-fs",
        "screen",
        "shell",
      ]
      .into_iter()
      .map(|i| ExternalItem::String(i.to_string()))
      .collect(),
      false,
    )
    .boxed(),
  );
  match context {
    ElectronTargetContext::Main => plugins.push(
      ExternalsPlugin::new(
        "node-commonjs".to_string(),
        [
          "app",
          "auto-updater",
          "browser-window",
          "content-tracing",
          "dialog",
          "global-shortcut",
          "ipc-main",
          "menu",
          "menu-item",
          "power-monitor",
          "power-save-blocker",
          "protocol",
          "session",
          "tray",
          "web-contents",
        ]
        .into_iter()
        .map(|i| ExternalItem::String(i.to_string()))
        .collect(),
        false,
      )
      .boxed(),
    ),
    ElectronTargetContext::Preload | ElectronTargetContext::Renderer => plugins.push(
      ExternalsPlugin::new(
        "node-commonjs".to_string(),
        ["desktop-capturer", "ipc-renderer", "remote", "web-frame"]
          .into_iter()
          .map(|i| ExternalItem::String(i.to_string()))
          .collect(),
        false,
      )
      .boxed(),
    ),
    ElectronTargetContext::None => {}
  }
}

File: packages/rspack/src/electron/ElectronTargetPlugin.ts
/**
 * The following code is modified based on
 * https://github.com/webpack/webpack/blob/4b4ca3b/lib/electron/ElectronTargetPlugin.js
 *
 * MIT Licensed
 * Author Tobias Koppers @sokra
 * Copyright (c) JS Foundation and other contributors
 * https://github.com/webpack/webpack/blob/main/LICENSE
 */

import { Compiler } from "..";
import { ExternalsPlugin } from "../builtin-plugin";

export default class ElectronTargetPlugin {
	constructor(private _context?: "main" | "preload" | "renderer") {}

	apply(compiler: Compiler) {
		new ExternalsPlugin("node-commonjs", [
			"clipboard",
			"crash-reporter",
			"electron",
			"ipc",
			"native-image",
			"original-fs",
			"screen",
			"shell"
		]).apply(compiler);
		switch (this._context) {
			case "main":
				new ExternalsPlugin("node-commonjs", [
					"app",
					"auto-updater",
					"browser-window",
					"content-tracing",
					"dialog",
					"global-shortcut",
					"ipc-main",
					"menu",
					"menu-item",
					"power-monitor",
					"power-save-blocker",
					"protocol",
					"session",
					"tray",
					"web-contents"
				]).apply(compiler);
				break;
			case "preload":
			case "renderer":
				new ExternalsPlugin("node-commonjs", [
					"desktop-capturer",
					"ipc-renderer",
					"remote",
					"web-frame"
				]).apply(compiler);
				break;
		}
	}
}
